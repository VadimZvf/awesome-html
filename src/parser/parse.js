import * as errors from './errors';

function isEmptyString(str) {
    return str.trim().length === 0;
}

export function getAttributesInfo(attributesString) {
    // TODO: добавить булевые аттрибуты
    // TODO: Добавить валидацию
    let matches;
    const attributes = [];

    const singleQuoteAttributeRegexp = /([\w-]+?)='([^'"<>]+?)'/gi;
    while ((matches = singleQuoteAttributeRegexp.exec(attributesString)) !== null) {
        attributes.push({ name: matches[1], value: matches[2] });
    }

    const doubleQuoteAttributeRegexp = /([\w-]+?)="([^'"<>]+?)"/gi;
    while ((matches = doubleQuoteAttributeRegexp.exec(attributesString)) !== null) {
        attributes.push({ name: matches[1], value: matches[2] });
    }

    return attributes;
}

export function getTagInfo(tagString, startIndex, endIndex) {
    const trimmedTagString = tagString.trim();

    let isCloseTag = /^\/.*$/.test(trimmedTagString); // </div>
    let isSelfClosed = /^.*\/$/.test(trimmedTagString); // <div/>
    let role = 'open';
    if (isCloseTag) {
        role = 'close';
    } else if (isSelfClosed) {
        role = 'open-close';
    }

    // TODO: добавить более подробные сорсы для тега
    const tagNameRegexp = /^\/?([^\s/]+).*?\/?$/gi;
    const tagNameMatch = tagNameRegexp.exec(trimmedTagString);
    const tagName = tagNameMatch ? tagNameMatch[1] : '';

    // </div/>
    if (isCloseTag && isSelfClosed && tagName) {
        const error = new Error(errors.INVALID_TAG_DECLARATION.getMessage());
        error.code = errors.INVALID_TAG_DECLARATION.code;
        error.source = { startIndex, endIndex };
        throw error;
    }

    const attributesString = trimmedTagString.replace(/^\/?([^\s/]+)/i, '').replace(/\/$/i, '');

    return {
        role,
        type: 'tag',
        name: tagName,
        attributes: getAttributesInfo(attributesString),
        source: { startIndex, endIndex }
    };
}

// Возврращает массив тегов,
// это еще не ноды, просто список найденных тегов или стоковых нод
export function getTags(inputValue = '') {
    if (!inputValue.length) {
        return [];
    }

    let currentText = '';
    let currentTagStart = null;
    const foundTagItems = [];

    for (let i = 0; i < inputValue.length; i++) {
        if (inputValue[i] === '<') {
            if (currentTagStart !== null) {
                // Начался новый тег, но предыдущий еще не закрыт ☹️
                const error = new Error(errors.WRONG_OPEN_TAG_SYMBOL.getMessage());
                error.code = errors.WRONG_OPEN_TAG_SYMBOL.code;
                error.source = { startIndex: i, endIndex: i };
                throw error;
            }

            // перед новым тегом была текстовая нода, надо закрыть её
            if (currentText && !isEmptyString(currentText)) {
                // проверяем, вдруг текстовая нода состояла только из пробелов
                foundTagItems.push({ type: 'text', text: currentText.trim() });
            }

            currentText = '';
            currentTagStart = i;
            continue;
        }

        if (inputValue[i] === '>') {
            if (currentTagStart === null) {
                // Перед символом закрытия тега не оказалось символа открытия тега
                const error = new Error(errors.WRONG_CLOSE_TAG_SYMBOL.getMessage());
                error.code = errors.WRONG_CLOSE_TAG_SYMBOL.code;
                error.source = { startIndex: i, endIndex: i };
                throw error;
            }

            foundTagItems.push(getTagInfo(currentText, currentTagStart, i));
            currentTagStart = null;
            currentText = '';
            continue;
        }

        // Мы находимся внутри тега или сейчас текстовая нода
        currentText = `${currentText}${inputValue[i]}`;
    }

    // прошлись по всей строке но текстовая нода не закрыта, значит на входе была просто строка
    if (currentText && !isEmptyString(currentText)) {
        // проверяем, вдруг текстовая нода состояла только из пробелов
        foundTagItems.push({ type: 'text', text: currentText });
    }

    // прошлись по всей строке но один тег не закрылся(
    if (currentTagStart !== null) {
        const error = new Error(errors.UNCLOSED_TAG.getMessage());
        error.code = errors.UNCLOSED_TAG.code;
        error.source = { startIndex: currentTagStart, endIndex: currentTagStart };
        throw error;
    }

    return foundTagItems;
}

function getPreparedTags(sourceTags) {
    return sourceTags.map((tag, index) => ({ ...tag, id: index, parentId: null, children: [] }));
}

export function getNodesMap(sourceTags) {
    const tags = getPreparedTags(sourceTags);
    const tagsMap = {};

    // делаем простой словарик, чтобы легко получать элемент
    for (const tag of tags) {
        // сохраняем только теги открытия, тк дубли не нужны
        if (tag.role === 'open') {
            tagsMap[tag.id] = tag;
        }

        if (tag.role === 'open-close') {
            tagsMap[tag.id] = tag;
        }

        if (tag.type === 'text') {
            tagsMap[tag.id] = tag;
        }
    }

    // список последних открытых тегов
    const currentOpenTags = [];

    for (let index = 0; index < tags.length; index++) {
        const currentTag = tags[index];
        const lastOpenTag = currentOpenTags[currentOpenTags.length - 1];

        if (currentTag.type === 'text') {
            if (lastOpenTag) {
                // Зачем текстовой ноде id родителя? О_о
                // будем искать по тексту?
                currentTag.parentId = lastOpenTag.id;
                lastOpenTag.children.push(currentTag);
            }
        }

        if (currentTag.role === 'open') {
            if (lastOpenTag) {
                currentTag.parentId = lastOpenTag.id;
                lastOpenTag.children.push(currentTag);
            }

            currentOpenTags.push(currentTag);
        }

        if (currentTag.role === 'open-close') {
            if (lastOpenTag) {
                currentTag.parentId = lastOpenTag.id;
                lastOpenTag.children.push(currentTag);
            }
        }

        if (currentTag.role === 'close') {
            if (!lastOpenTag || lastOpenTag.name !== currentTag.name) {
                const error = new Error(errors.WRONG_CLOSE_TAG.getMessage(currentTag));
                error.code = errors.WRONG_CLOSE_TAG.code;
                error.source = currentTag.source;

                throw error;
            }

            currentOpenTags.pop();
            // prevOpenTag
        }
    }

    if (currentOpenTags.length) {
        const notClosedTag = currentOpenTags[currentOpenTags.length - 1];
        const error = new Error(errors.WRONG_OPEN_TAG.getMessage(notClosedTag));
        error.code = errors.WRONG_OPEN_TAG.code;
        error.source = notClosedTag.source;

        throw error;
    }

    return tagsMap;
}

export function getNodesTree(map) {
    return Object.values(map).find(node => !node.parentId);
}

// возвращает готовую AST
function parse(sourceString) {
    const tags = getTags(sourceString);
    const map = getNodesMap(tags);
    const tree = getNodesTree(map);

    return {
        tree,
        map // для простого поиска
    };
}

export default parse;
