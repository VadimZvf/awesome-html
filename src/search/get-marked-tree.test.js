import parse from '../parser/parse';
import getMarkedTree from './get-marked-tree';

const divNode = { name: 'div', type: 'tag', role: 'open', attributes: [], children: [] };
const htmlNode = { name: 'html', type: 'tag', role: 'open', attributes: [], children: [] };
// const textNode = { type: 'text', text: 'some text' };

const classSelector = { type: 'attribute', name: 'class', value: 'classSelector' };
const idSelector = { type: 'attribute', name: 'id', value: 'foo' };
const tagSelector = { type: 'tag', name: null, value: 'div' };

describe('getMarkedTree', () => {
    test('Should mark simple tree', () => {
        /**
         * <div>
         *    <div class="classSelector"></div>
         * </div>
         */

        const targetDivNode = {
            ...divNode,
            id: 1,
            parentId: 0,
            attributes: [{ name: 'class', value: 'classSelector' }]
        };
        const map = {
            0: { ...divNode, id: 0, parentId: null, children: [targetDivNode] },
            1: targetDivNode
        };

        expect(getMarkedTree(map, [classSelector])).toStrictEqual({
            tree: {
                ...divNode,
                id: 0,
                parentId: null,
                isFound: false,
                isOpen: true,
                children: [{ ...targetDivNode, isOpen: false, isFound: true }]
            },
            map: {
                0: {
                    ...divNode,
                    id: 0,
                    parentId: null,
                    isFound: false,
                    isOpen: true,
                    children: [{ ...targetDivNode, isOpen: false, isFound: true }]
                },
                1: { ...targetDivNode, isOpen: false, isFound: true }
            }
        });
    });
    test('Should search by two selectors', () => {
        // ".bar .baz"
        const query = [
            { ...classSelector, value: 'bar' },
            { ...classSelector, value: 'baz' }
        ];

        const bazDivNode = {
            ...divNode,
            id: 2,
            parentId: 1,
            isFound: false,
            isOpen: false,
            source: { endIndex: 75, startIndex: 59 },
            attributes: [{ name: 'class', value: 'baz' }]
        };
        const fooDivNode = {
            ...divNode,
            id: 1,
            parentId: 0,
            isFound: false,
            isOpen: false,
            source: { endIndex: 42, startIndex: 26 },
            attributes: [{ name: 'class', value: 'foo' }],
            children: [bazDivNode]
        };
        const secondBazDivNode = {
            ...divNode,
            id: 6,
            parentId: 5,
            isFound: true,
            isOpen: false,
            source: { endIndex: 161, startIndex: 145 },
            attributes: [{ name: 'class', value: 'baz' }]
        };
        const barDivNode = {
            ...divNode,
            id: 5,
            parentId: 0,
            isFound: false,
            isOpen: true,
            source: { endIndex: 128, startIndex: 112 },
            attributes: [{ name: 'class', value: 'bar' }],
            children: [secondBazDivNode]
        };
        const rootNode = {
            ...divNode,
            id: 0,
            parentId: null,
            isFound: false,
            isOpen: true,
            source: { endIndex: 13, startIndex: 9 },
            children: [fooDivNode, barDivNode]
        };

        const { map } = parse(`
        <div>
           <div class="foo">
               <div class="baz"></div>
           </div>
           <div class="bar">
               <div class="baz"></div>
           </div>
        </div>
        `);

        expect(getMarkedTree(map, query)).toStrictEqual({
            tree: rootNode,
            map: {
                0: rootNode,
                1: fooDivNode,
                2: bazDivNode,
                5: barDivNode,
                6: secondBazDivNode
            }
        });
    });

    test('Should search by tag', () => {
        // "div"
        const query = [tagSelector];

        const fooDivNode = {
            ...divNode,
            id: 1,
            parentId: 0,
            isFound: true,
            isOpen: false,
            source: { endIndex: 32, startIndex: 28 },
            attributes: [],
            children: []
        };
        const rootNode = {
            ...htmlNode,
            id: 0,
            parentId: null,
            isFound: false,
            isOpen: true,
            source: { endIndex: 14, startIndex: 9 },
            children: [fooDivNode]
        };

        const { map } = parse(`
        <html>
            <div></div>
        </html>
        `);

        expect(getMarkedTree(map, query)).toStrictEqual({
            tree: rootNode,
            map: {
                0: rootNode,
                1: fooDivNode
            }
        });
    });

    test('Should search to deep, skip not matched nodes', () => {
        // ".foo .bar"
        const query = [
            { ...classSelector, value: 'foo' },
            { ...classSelector, value: 'bar' }
        ];

        const bazDivNode = {
            ...divNode,
            id: 4,
            parentId: 1,
            isFound: false,
            isOpen: false,
            source: { endIndex: 118, startIndex: 102 },
            attributes: [{ name: 'class', value: 'baz' }]
        };
        const barDivNode = {
            ...divNode,
            id: 2,
            parentId: 1,
            isFound: true,
            isOpen: false,
            source: { endIndex: 78, startIndex: 62 },
            attributes: [{ name: 'class', value: 'bar' }],
            children: []
        };
        const divWrapNode = {
            ...divNode,
            id: 1,
            parentId: 0,
            isFound: false,
            isOpen: true,
            source: { endIndex: 44, startIndex: 40 },
            children: [barDivNode, bazDivNode]
        };
        const rootNode = {
            ...htmlNode,
            id: 0,
            parentId: null,
            isFound: false,
            isOpen: true,
            attributes: [{ name: 'class', value: 'foo' }],
            source: { endIndex: 26, startIndex: 9 },
            children: [divWrapNode]
        };

        const { map } = parse(`
        <html class="foo">
            <div>
                <div class="bar"></div>
                <div class="baz"></div>
            </div>
        </html>
        `);

        expect(getMarkedTree(map, query)).toStrictEqual({
            tree: rootNode,
            map: {
                0: rootNode,
                1: divWrapNode,
                2: barDivNode,
                4: bazDivNode
            }
        });
    });

    test('Should only one node by id', () => {
        // "#foo"
        const query = [idSelector];

        const fooDivNode = {
            ...divNode,
            id: 1,
            parentId: 0,
            isFound: true,
            isOpen: false,
            source: { endIndex: 48, startIndex: 35 },
            attributes: [{ name: 'id', value: 'foo' }]
        };
        const fooSecondDivNode = {
            ...divNode,
            id: 3,
            parentId: 0,
            isFound: false,
            isOpen: false,
            source: { endIndex: 85, startIndex: 72 },
            attributes: [{ name: 'id', value: 'foo' }],
            children: []
        };
        const rootNode = {
            ...divNode,
            id: 0,
            parentId: null,
            isFound: false,
            isOpen: true,
            attributes: [],
            source: { endIndex: 17, startIndex: 13 },
            children: [fooDivNode, fooSecondDivNode]
        };

        const { map } = parse(`
            <div>
                <div id="foo"></div>
                <div id="foo"></div>
            </div>
        `);

        expect(getMarkedTree(map, query)).toStrictEqual({
            tree: rootNode,
            map: {
                0: rootNode,
                1: fooDivNode,
                3: fooSecondDivNode
            }
        });
    });
});
