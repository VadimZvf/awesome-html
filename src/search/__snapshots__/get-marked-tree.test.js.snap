// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getMarkedTree Should find first child node 1`] = `
Object {
  "map": Object {
    "0": Object {
      "attributes": Array [
        Object {
          "name": "id",
          "value": "foo",
        },
      ],
      "children": Array [
        Object {
          "attributes": Array [
            Object {
              "name": "class",
              "value": "bar",
            },
          ],
          "children": Array [],
          "id": 1,
          "isFound": true,
          "isOpen": false,
          "name": "div",
          "parentId": 0,
          "role": "open",
          "source": Object {
            "endIndex": 60,
            "startIndex": 44,
          },
          "type": "tag",
        },
        Object {
          "attributes": Array [],
          "children": Array [
            Object {
              "attributes": Array [
                Object {
                  "name": "class",
                  "value": "bar",
                },
              ],
              "children": Array [],
              "id": 4,
              "isFound": false,
              "isOpen": false,
              "name": "div",
              "parentId": 3,
              "role": "open",
              "source": Object {
                "endIndex": 126,
                "startIndex": 110,
              },
              "type": "tag",
            },
          ],
          "id": 3,
          "isFound": false,
          "isOpen": false,
          "name": "div",
          "parentId": 0,
          "role": "open",
          "source": Object {
            "endIndex": 88,
            "startIndex": 84,
          },
          "type": "tag",
        },
      ],
      "id": 0,
      "isFound": false,
      "isOpen": true,
      "name": "div",
      "parentId": null,
      "role": "open",
      "source": Object {
        "endIndex": 26,
        "startIndex": 13,
      },
      "type": "tag",
    },
    "1": Object {
      "attributes": Array [
        Object {
          "name": "class",
          "value": "bar",
        },
      ],
      "children": Array [],
      "id": 1,
      "isFound": true,
      "isOpen": false,
      "name": "div",
      "parentId": 0,
      "role": "open",
      "source": Object {
        "endIndex": 60,
        "startIndex": 44,
      },
      "type": "tag",
    },
    "3": Object {
      "attributes": Array [],
      "children": Array [
        Object {
          "attributes": Array [
            Object {
              "name": "class",
              "value": "bar",
            },
          ],
          "children": Array [],
          "id": 4,
          "isFound": false,
          "isOpen": false,
          "name": "div",
          "parentId": 3,
          "role": "open",
          "source": Object {
            "endIndex": 126,
            "startIndex": 110,
          },
          "type": "tag",
        },
      ],
      "id": 3,
      "isFound": false,
      "isOpen": false,
      "name": "div",
      "parentId": 0,
      "role": "open",
      "source": Object {
        "endIndex": 88,
        "startIndex": 84,
      },
      "type": "tag",
    },
    "4": Object {
      "attributes": Array [
        Object {
          "name": "class",
          "value": "bar",
        },
      ],
      "children": Array [],
      "id": 4,
      "isFound": false,
      "isOpen": false,
      "name": "div",
      "parentId": 3,
      "role": "open",
      "source": Object {
        "endIndex": 126,
        "startIndex": 110,
      },
      "type": "tag",
    },
  },
  "tree": Object {
    "attributes": Array [
      Object {
        "name": "id",
        "value": "foo",
      },
    ],
    "children": Array [
      Object {
        "attributes": Array [
          Object {
            "name": "class",
            "value": "bar",
          },
        ],
        "children": Array [],
        "id": 1,
        "isFound": true,
        "isOpen": false,
        "name": "div",
        "parentId": 0,
        "role": "open",
        "source": Object {
          "endIndex": 60,
          "startIndex": 44,
        },
        "type": "tag",
      },
      Object {
        "attributes": Array [],
        "children": Array [
          Object {
            "attributes": Array [
              Object {
                "name": "class",
                "value": "bar",
              },
            ],
            "children": Array [],
            "id": 4,
            "isFound": false,
            "isOpen": false,
            "name": "div",
            "parentId": 3,
            "role": "open",
            "source": Object {
              "endIndex": 126,
              "startIndex": 110,
            },
            "type": "tag",
          },
        ],
        "id": 3,
        "isFound": false,
        "isOpen": false,
        "name": "div",
        "parentId": 0,
        "role": "open",
        "source": Object {
          "endIndex": 88,
          "startIndex": 84,
        },
        "type": "tag",
      },
    ],
    "id": 0,
    "isFound": false,
    "isOpen": true,
    "name": "div",
    "parentId": null,
    "role": "open",
    "source": Object {
      "endIndex": 26,
      "startIndex": 13,
    },
    "type": "tag",
  },
}
`;

exports[`getMarkedTree Should find next node 1`] = `
Object {
  "map": Object {
    "0": Object {
      "attributes": Array [],
      "children": Array [
        Object {
          "attributes": Array [
            Object {
              "name": "id",
              "value": "foo",
            },
          ],
          "children": Array [],
          "id": 1,
          "isFound": false,
          "isOpen": false,
          "name": "div",
          "parentId": 0,
          "role": "open",
          "source": Object {
            "endIndex": 48,
            "startIndex": 35,
          },
          "type": "tag",
        },
        Object {
          "attributes": Array [
            Object {
              "name": "class",
              "value": "bar",
            },
          ],
          "children": Array [
            Object {
              "attributes": Array [],
              "children": Array [],
              "id": 4,
              "isFound": false,
              "isOpen": false,
              "name": "div",
              "parentId": 3,
              "role": "open",
              "source": Object {
                "endIndex": 114,
                "startIndex": 110,
              },
              "type": "tag",
            },
          ],
          "id": 3,
          "isFound": true,
          "isOpen": false,
          "name": "div",
          "parentId": 0,
          "role": "open",
          "source": Object {
            "endIndex": 88,
            "startIndex": 72,
          },
          "type": "tag",
        },
      ],
      "id": 0,
      "isFound": false,
      "isOpen": true,
      "name": "div",
      "parentId": null,
      "role": "open",
      "source": Object {
        "endIndex": 17,
        "startIndex": 13,
      },
      "type": "tag",
    },
    "1": Object {
      "attributes": Array [
        Object {
          "name": "id",
          "value": "foo",
        },
      ],
      "children": Array [],
      "id": 1,
      "isFound": false,
      "isOpen": false,
      "name": "div",
      "parentId": 0,
      "role": "open",
      "source": Object {
        "endIndex": 48,
        "startIndex": 35,
      },
      "type": "tag",
    },
    "3": Object {
      "attributes": Array [
        Object {
          "name": "class",
          "value": "bar",
        },
      ],
      "children": Array [
        Object {
          "attributes": Array [],
          "children": Array [],
          "id": 4,
          "isFound": false,
          "isOpen": false,
          "name": "div",
          "parentId": 3,
          "role": "open",
          "source": Object {
            "endIndex": 114,
            "startIndex": 110,
          },
          "type": "tag",
        },
      ],
      "id": 3,
      "isFound": true,
      "isOpen": false,
      "name": "div",
      "parentId": 0,
      "role": "open",
      "source": Object {
        "endIndex": 88,
        "startIndex": 72,
      },
      "type": "tag",
    },
    "4": Object {
      "attributes": Array [],
      "children": Array [],
      "id": 4,
      "isFound": false,
      "isOpen": false,
      "name": "div",
      "parentId": 3,
      "role": "open",
      "source": Object {
        "endIndex": 114,
        "startIndex": 110,
      },
      "type": "tag",
    },
  },
  "tree": Object {
    "attributes": Array [],
    "children": Array [
      Object {
        "attributes": Array [
          Object {
            "name": "id",
            "value": "foo",
          },
        ],
        "children": Array [],
        "id": 1,
        "isFound": false,
        "isOpen": false,
        "name": "div",
        "parentId": 0,
        "role": "open",
        "source": Object {
          "endIndex": 48,
          "startIndex": 35,
        },
        "type": "tag",
      },
      Object {
        "attributes": Array [
          Object {
            "name": "class",
            "value": "bar",
          },
        ],
        "children": Array [
          Object {
            "attributes": Array [],
            "children": Array [],
            "id": 4,
            "isFound": false,
            "isOpen": false,
            "name": "div",
            "parentId": 3,
            "role": "open",
            "source": Object {
              "endIndex": 114,
              "startIndex": 110,
            },
            "type": "tag",
          },
        ],
        "id": 3,
        "isFound": true,
        "isOpen": false,
        "name": "div",
        "parentId": 0,
        "role": "open",
        "source": Object {
          "endIndex": 88,
          "startIndex": 72,
        },
        "type": "tag",
      },
    ],
    "id": 0,
    "isFound": false,
    "isOpen": true,
    "name": "div",
    "parentId": null,
    "role": "open",
    "source": Object {
      "endIndex": 17,
      "startIndex": 13,
    },
    "type": "tag",
  },
}
`;

exports[`getMarkedTree Should find next node, that matched by query 1`] = `
Object {
  "map": Object {
    "0": Object {
      "attributes": Array [],
      "children": Array [
        Object {
          "attributes": Array [
            Object {
              "name": "id",
              "value": "foo",
            },
          ],
          "children": Array [],
          "id": 1,
          "isFound": false,
          "isOpen": false,
          "name": "div",
          "parentId": 0,
          "role": "open",
          "source": Object {
            "endIndex": 48,
            "startIndex": 35,
          },
          "type": "tag",
        },
        Object {
          "attributes": Array [],
          "children": Array [],
          "id": 3,
          "isFound": false,
          "isOpen": false,
          "name": "div",
          "parentId": 0,
          "role": "open",
          "source": Object {
            "endIndex": 76,
            "startIndex": 72,
          },
          "type": "tag",
        },
        Object {
          "attributes": Array [
            Object {
              "name": "class",
              "value": "bar",
            },
          ],
          "children": Array [],
          "id": 5,
          "isFound": true,
          "isOpen": false,
          "name": "div",
          "parentId": 0,
          "role": "open",
          "source": Object {
            "endIndex": 116,
            "startIndex": 100,
          },
          "type": "tag",
        },
      ],
      "id": 0,
      "isFound": false,
      "isOpen": true,
      "name": "div",
      "parentId": null,
      "role": "open",
      "source": Object {
        "endIndex": 17,
        "startIndex": 13,
      },
      "type": "tag",
    },
    "1": Object {
      "attributes": Array [
        Object {
          "name": "id",
          "value": "foo",
        },
      ],
      "children": Array [],
      "id": 1,
      "isFound": false,
      "isOpen": false,
      "name": "div",
      "parentId": 0,
      "role": "open",
      "source": Object {
        "endIndex": 48,
        "startIndex": 35,
      },
      "type": "tag",
    },
    "3": Object {
      "attributes": Array [],
      "children": Array [],
      "id": 3,
      "isFound": false,
      "isOpen": false,
      "name": "div",
      "parentId": 0,
      "role": "open",
      "source": Object {
        "endIndex": 76,
        "startIndex": 72,
      },
      "type": "tag",
    },
    "5": Object {
      "attributes": Array [
        Object {
          "name": "class",
          "value": "bar",
        },
      ],
      "children": Array [],
      "id": 5,
      "isFound": true,
      "isOpen": false,
      "name": "div",
      "parentId": 0,
      "role": "open",
      "source": Object {
        "endIndex": 116,
        "startIndex": 100,
      },
      "type": "tag",
    },
  },
  "tree": Object {
    "attributes": Array [],
    "children": Array [
      Object {
        "attributes": Array [
          Object {
            "name": "id",
            "value": "foo",
          },
        ],
        "children": Array [],
        "id": 1,
        "isFound": false,
        "isOpen": false,
        "name": "div",
        "parentId": 0,
        "role": "open",
        "source": Object {
          "endIndex": 48,
          "startIndex": 35,
        },
        "type": "tag",
      },
      Object {
        "attributes": Array [],
        "children": Array [],
        "id": 3,
        "isFound": false,
        "isOpen": false,
        "name": "div",
        "parentId": 0,
        "role": "open",
        "source": Object {
          "endIndex": 76,
          "startIndex": 72,
        },
        "type": "tag",
      },
      Object {
        "attributes": Array [
          Object {
            "name": "class",
            "value": "bar",
          },
        ],
        "children": Array [],
        "id": 5,
        "isFound": true,
        "isOpen": false,
        "name": "div",
        "parentId": 0,
        "role": "open",
        "source": Object {
          "endIndex": 116,
          "startIndex": 100,
        },
        "type": "tag",
      },
    ],
    "id": 0,
    "isFound": false,
    "isOpen": true,
    "name": "div",
    "parentId": null,
    "role": "open",
    "source": Object {
      "endIndex": 17,
      "startIndex": 13,
    },
    "type": "tag",
  },
}
`;

exports[`getMarkedTree Should find only one element by id and class 1`] = `
Object {
  "map": Object {
    "0": Object {
      "attributes": Array [],
      "children": Array [
        Object {
          "attributes": Array [
            Object {
              "name": "id",
              "value": "foo",
            },
          ],
          "children": Array [],
          "id": 1,
          "isFound": false,
          "isOpen": false,
          "name": "div",
          "parentId": 0,
          "role": "open",
          "source": Object {
            "endIndex": 48,
            "startIndex": 35,
          },
          "type": "tag",
        },
        Object {
          "attributes": Array [
            Object {
              "name": "class",
              "value": "bar",
            },
          ],
          "children": Array [
            Object {
              "attributes": Array [
                Object {
                  "name": "id",
                  "value": "foo",
                },
              ],
              "children": Array [],
              "id": 4,
              "isFound": true,
              "isOpen": false,
              "name": "div",
              "parentId": 3,
              "role": "open",
              "source": Object {
                "endIndex": 123,
                "startIndex": 110,
              },
              "type": "tag",
            },
          ],
          "id": 3,
          "isFound": false,
          "isOpen": true,
          "name": "div",
          "parentId": 0,
          "role": "open",
          "source": Object {
            "endIndex": 88,
            "startIndex": 72,
          },
          "type": "tag",
        },
      ],
      "id": 0,
      "isFound": false,
      "isOpen": true,
      "name": "div",
      "parentId": null,
      "role": "open",
      "source": Object {
        "endIndex": 17,
        "startIndex": 13,
      },
      "type": "tag",
    },
    "1": Object {
      "attributes": Array [
        Object {
          "name": "id",
          "value": "foo",
        },
      ],
      "children": Array [],
      "id": 1,
      "isFound": false,
      "isOpen": false,
      "name": "div",
      "parentId": 0,
      "role": "open",
      "source": Object {
        "endIndex": 48,
        "startIndex": 35,
      },
      "type": "tag",
    },
    "3": Object {
      "attributes": Array [
        Object {
          "name": "class",
          "value": "bar",
        },
      ],
      "children": Array [
        Object {
          "attributes": Array [
            Object {
              "name": "id",
              "value": "foo",
            },
          ],
          "children": Array [],
          "id": 4,
          "isFound": true,
          "isOpen": false,
          "name": "div",
          "parentId": 3,
          "role": "open",
          "source": Object {
            "endIndex": 123,
            "startIndex": 110,
          },
          "type": "tag",
        },
      ],
      "id": 3,
      "isFound": false,
      "isOpen": true,
      "name": "div",
      "parentId": 0,
      "role": "open",
      "source": Object {
        "endIndex": 88,
        "startIndex": 72,
      },
      "type": "tag",
    },
    "4": Object {
      "attributes": Array [
        Object {
          "name": "id",
          "value": "foo",
        },
      ],
      "children": Array [],
      "id": 4,
      "isFound": true,
      "isOpen": false,
      "name": "div",
      "parentId": 3,
      "role": "open",
      "source": Object {
        "endIndex": 123,
        "startIndex": 110,
      },
      "type": "tag",
    },
  },
  "tree": Object {
    "attributes": Array [],
    "children": Array [
      Object {
        "attributes": Array [
          Object {
            "name": "id",
            "value": "foo",
          },
        ],
        "children": Array [],
        "id": 1,
        "isFound": false,
        "isOpen": false,
        "name": "div",
        "parentId": 0,
        "role": "open",
        "source": Object {
          "endIndex": 48,
          "startIndex": 35,
        },
        "type": "tag",
      },
      Object {
        "attributes": Array [
          Object {
            "name": "class",
            "value": "bar",
          },
        ],
        "children": Array [
          Object {
            "attributes": Array [
              Object {
                "name": "id",
                "value": "foo",
              },
            ],
            "children": Array [],
            "id": 4,
            "isFound": true,
            "isOpen": false,
            "name": "div",
            "parentId": 3,
            "role": "open",
            "source": Object {
              "endIndex": 123,
              "startIndex": 110,
            },
            "type": "tag",
          },
        ],
        "id": 3,
        "isFound": false,
        "isOpen": true,
        "name": "div",
        "parentId": 0,
        "role": "open",
        "source": Object {
          "endIndex": 88,
          "startIndex": 72,
        },
        "type": "tag",
      },
    ],
    "id": 0,
    "isFound": false,
    "isOpen": true,
    "name": "div",
    "parentId": null,
    "role": "open",
    "source": Object {
      "endIndex": 17,
      "startIndex": 13,
    },
    "type": "tag",
  },
}
`;

exports[`getMarkedTree Should search node with two classes 1`] = `
Object {
  "map": Object {
    "0": Object {
      "attributes": Array [
        Object {
          "name": "class",
          "value": "foo",
        },
      ],
      "children": Array [
        Object {
          "attributes": Array [
            Object {
              "name": "class",
              "value": "bar",
            },
          ],
          "children": Array [],
          "id": 1,
          "isFound": false,
          "isOpen": false,
          "name": "div",
          "parentId": 0,
          "role": "open",
          "source": Object {
            "endIndex": 63,
            "startIndex": 47,
          },
          "type": "tag",
        },
        Object {
          "attributes": Array [
            Object {
              "name": "class",
              "value": "foo bar",
            },
          ],
          "children": Array [],
          "id": 3,
          "isFound": true,
          "isOpen": false,
          "name": "div",
          "parentId": 0,
          "role": "open",
          "source": Object {
            "endIndex": 107,
            "startIndex": 87,
          },
          "type": "tag",
        },
      ],
      "id": 0,
      "isFound": false,
      "isOpen": true,
      "name": "div",
      "parentId": null,
      "role": "open",
      "source": Object {
        "endIndex": 29,
        "startIndex": 13,
      },
      "type": "tag",
    },
    "1": Object {
      "attributes": Array [
        Object {
          "name": "class",
          "value": "bar",
        },
      ],
      "children": Array [],
      "id": 1,
      "isFound": false,
      "isOpen": false,
      "name": "div",
      "parentId": 0,
      "role": "open",
      "source": Object {
        "endIndex": 63,
        "startIndex": 47,
      },
      "type": "tag",
    },
    "3": Object {
      "attributes": Array [
        Object {
          "name": "class",
          "value": "foo bar",
        },
      ],
      "children": Array [],
      "id": 3,
      "isFound": true,
      "isOpen": false,
      "name": "div",
      "parentId": 0,
      "role": "open",
      "source": Object {
        "endIndex": 107,
        "startIndex": 87,
      },
      "type": "tag",
    },
  },
  "tree": Object {
    "attributes": Array [
      Object {
        "name": "class",
        "value": "foo",
      },
    ],
    "children": Array [
      Object {
        "attributes": Array [
          Object {
            "name": "class",
            "value": "bar",
          },
        ],
        "children": Array [],
        "id": 1,
        "isFound": false,
        "isOpen": false,
        "name": "div",
        "parentId": 0,
        "role": "open",
        "source": Object {
          "endIndex": 63,
          "startIndex": 47,
        },
        "type": "tag",
      },
      Object {
        "attributes": Array [
          Object {
            "name": "class",
            "value": "foo bar",
          },
        ],
        "children": Array [],
        "id": 3,
        "isFound": true,
        "isOpen": false,
        "name": "div",
        "parentId": 0,
        "role": "open",
        "source": Object {
          "endIndex": 107,
          "startIndex": 87,
        },
        "type": "tag",
      },
    ],
    "id": 0,
    "isFound": false,
    "isOpen": true,
    "name": "div",
    "parentId": null,
    "role": "open",
    "source": Object {
      "endIndex": 29,
      "startIndex": 13,
    },
    "type": "tag",
  },
}
`;
